### PR TITLE
reconcile master service if it becomes incorrect

### LINKED ISSUE #16225: After master upgrade (v1.0->HEAD) DNS doesn't work because default service doesn't work
Looks like we added a name to the kubernetes service port, but we only update the endpoints with the new port name and not the service. This trips up the kube-proxy because it can't find a svcportname mapping from the service to the endpoints.

cc @davidopp @thockin @ArtfulCoder @ihmccreery

### PR DESCRIPTION
**Pull request by mikedanese**
I'm not completely sold on this approach. It may become an issue when apiserver is replicated. API Servers of different versions could potentially war over the hash value. On the other hand, maybe this controller will have been completely rewritten by that time. We also have better control over master upgrades.

cc @thockin 

Fixes #16225

### CODE REVIEW

File: pkg/master/controller.go [hunk 1]
```diff
@@ -200,6 +197,19 @@ func (c *Controller) CreateMasterServiceIfNeeded(serviceName string, serviceIP n
 			Type:            serviceType,
 		},
 	}
+	hash := strconv.FormatUint(hashService(svc), 10)
+	svc.Annotations = map[string]string{"kubernetes.io/config.hash": hash}
+	ctx := api.NewDefaultContext()
+	if curSvc, err := c.ServiceRegistry.GetService(ctx, serviceName); err == nil {
+		if curHash, found := curSvc.Annotations["kubernetes.io/config.hash"]; found && hash == curHash {
```
Comment (lavalamp) [inline pkg/master/controller.go]: Why would you actually store the hash? Shouldn't you recompute it every time?

File: pkg/master/controller.go [hunk 2]
```diff
@@ -200,6 +197,19 @@ func (c *Controller) CreateMasterServiceIfNeeded(serviceName string, serviceIP n
 			Type:            serviceType,
 		},
 	}
+	hash := strconv.FormatUint(hashService(svc), 10)
+	svc.Annotations = map[string]string{"kubernetes.io/config.hash": hash}
```
Comment (lavalamp) [inline pkg/master/controller.go]: Putting the hash in the annotations makes the hash wrong, unless you want to do something like bitcoin mining to make it not wrong ;)

Just saying...

File: pkg/master/controller.go (hunk 1 above)
Comment (lavalamp) [inline pkg/master/controller.go]: I guess this is to be tolerant to defaults getting filled in?

File: pkg/master/controller.go (hunk 2 above)
Comment (lavalamp) [inline pkg/master/controller.go]: Also @bgrant0607 will be happier if you change that annotation key to "apiserver.kubernetes.io/master-service-hash" or some such.

**Comment by lavalamp** [comment]
OK so the issue with this is that you can't do a rolling upgrade of masters if this changes, since the old ones and the new ones will fight.

**Comment by bgrant0607** [comment]
@lavalamp is correct re. annotation convention.

Annotation conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#label-selector-and-annotation-conventions

Also, please don't reuse the annotation that Kubelet is using, since that would be confusing later.

**Comment by mikedanese** [comment]
Might have been superseeded by #15791

**Pull request closed without merging**